demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+    size_t fileSize = S3_FILE_SIZE;
+
+    /* The number of bytes we want to request within each range of the file. */
+    size_t numReqBytes = 0;
+    /* curByte indicates which starting byte we want to download next. */
+    size_t curByte = 0;
+
+    /* Initialize the request object. */
+    requestInfo.pHost = pHost;
+    requestInfo.hostLen = strlen( pHost );
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = strlen( pMethod );
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. This is done in
+     * order to download the file in parts. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestItem.requestHeaders.pBuffer = requestItem.headersBuffer;
+    requestItem.requestHeaders.bufferLen = sizeof( requestItem.headersBuffer );
+
+    /* Ensure requested amount will fit in buffer */
+    if( fileSize < sizeof( requestItem.headersBuffer ) - HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES )
+    {
+        numReqBytes = fileSize;
+    }
+    else
+    {
+        numReqBytes = sizeof( requestItem.headersBuffer ) - HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES;
+    }
+
+    /* Here we iterate sending byte range requests until the full file has been
+     * downloaded. We keep track of the next byte to download with curByte.
+     * When this reaches the fileSize we stop downloading. */
+    while( curByte < fileSize && httpStatus == HTTP_SUCCESS && returnStatus == EXIT_SUCCESS )
+    {
+        if( httpStatus == HTTP_SUCCESS )
+        {
+            httpStatus = HTTPClient_InitializeRequestHeaders( &( requestItem.requestHeaders ),
+                                                              &requestInfo );
+        }
+
+        if( httpStatus == HTTP_SUCCESS )
+        {
+            httpStatus = HTTPClient_AddRangeHeader( &( requestItem.requestHeaders ),
+                                                    curByte,
+                                                    curByte + numReqBytes - 1 );
+        }
+        else
+        {
+            LogError( ( "Failed to initialize HTTP request headers: Error=%s.",
+                        HTTPClient_strerror( httpStatus ) ) );
+        }
+
+        if( httpStatus == HTTP_SUCCESS )
+        {
+            /* Enqueue the request. */
+            LogInfo( ( "Enqueuing %d bytes of S3 Object out of %d total bytes, from %s...:  ",
+                       ( int32_t ) ( curByte + numReqBytes - 1 ),
+                       ( int32_t ) fileSize,
+                       pHost ) );
+            LogInfo( ( "Request Headers:\n%.*s",
+                       ( int32_t ) requestItem.requestHeaders.headersLen,
+                       ( char * ) requestItem.requestHeaders.pBuffer ) );
+
+            mqerror = mq_send( requestQueue,
```

Comment:
So you have the one local requestItem on line 268 that is being put into this queue and you will keep adding more items to the queue for each of the blocks. Does this function allocate memory for each of the requestItems that goes into the queue?